File: pgpainless-core/src/main/java/org/pgpainless/algorithm/PublicKeyAlgorithm.java
// SPDX-FileCopyrightText: 2018 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.algorithm;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.bouncycastle.bcpg.PublicKeyAlgorithmTags;

/**
 * Enumeration of public key algorithms as defined in RFC4880.
 *
 * @see <a href="https://tools.ietf.org/html/rfc4880#section-9.1">RFC4880: Public-Key Algorithms</a>
 */
public enum PublicKeyAlgorithm {

    /**
     * RSA capable of encryption and signatures.
     */
    RSA_GENERAL     (PublicKeyAlgorithmTags.RSA_GENERAL, true, true),

    /**
     * RSA with usage encryption.
     *
     * @deprecated see https://tools.ietf.org/html/rfc4880#section-13.5
     */
    @Deprecated
    RSA_ENCRYPT     (PublicKeyAlgorithmTags.RSA_ENCRYPT, false, true),

    /**
     * RSA with usage of creating signatures.
     *
     * @deprecated see https://tools.ietf.org/html/rfc4880#section-13.5
     */
    @Deprecated
    RSA_SIGN        (PublicKeyAlgorithmTags.RSA_SIGN, true, false),

    /**
     * ElGamal with usage encryption.
     */
    ELGAMAL_ENCRYPT (PublicKeyAlgorithmTags.ELGAMAL_ENCRYPT, false, true),

    /**
     * Digital Signature Algorithm.
     */
    DSA             (PublicKeyAlgorithmTags.DSA, true, false),

    /**
     * EC is deprecated.
     * @deprecated use {@link #ECDH} instead.
     */
    @Deprecated
    EC              (PublicKeyAlgorithmTags.EC, false, true),

    /**
     * Elliptic Curve Diffie-Hellman.
     */
    ECDH            (PublicKeyAlgorithmTags.ECDH, false, true),

    /**
     * Elliptic Curve Digital Signature Algorithm.
     */
    ECDSA           (PublicKeyAlgorithmTags.ECDSA, true, false),

    /**
     * ElGamal General.
     *
     * @deprecated see https://tools.ietf.org/html/rfc4880#section-13.8
     */
    @Deprecated
    ELGAMAL_GENERAL (PublicKeyAlgorithmTags.ELGAMAL_GENERAL, true, true),

    /**
     * Diffie-Hellman key exchange algorithm.
     */
    DIFFIE_HELLMAN  (PublicKeyAlgorithmTags.DIFFIE_HELLMAN, false, true),

    /**
     * Digital Signature Algorithm based on twisted Edwards Curves.
     */
    EDDSA           (PublicKeyAlgorithmTags.EDDSA, true, false),
    ;

    private static final Map<Integer, PublicKeyAlgorithm> MAP = new ConcurrentHashMap<>();

    static {
        for (PublicKeyAlgorithm p : PublicKeyAlgorithm.values()) {
            MAP.put(p.algorithmId, p);
        }
    }

    /**
     * Return the {@link PublicKeyAlgorithm} that corresponds to the provided algorithm id.
     * If an invalid id is provided, null is returned.
     *
     * @param id numeric algorithm id
     * @return algorithm
     */
    public static PublicKeyAlgorithm fromId(int id) {
        return MAP.get(id);
    }

    private final int algorithmId;
    private final boolean signingCapable;
    private final boolean encryptionCapable;

    PublicKeyAlgorithm(int algorithmId, boolean signingCapable, boolean encryptionCapable) {
        this.algorithmId = algorithmId;
        this.signingCapable = signingCapable;
        this.encryptionCapable = encryptionCapable;
    }

    /**
     * Return the numeric identifier of the public key algorithm.
     *
     * @return id
     */
    public int getAlgorithmId() {
        return algorithmId;
    }

    /**
     * Return true if this public key algorithm is able to create signatures.
     *
     * @return true if the algorithm can sign
     */
    public boolean isSigningCapable() {
        return signingCapable;
    }

    /**
     * Return true if this public key algorithm can be used as an encryption algorithm.
     *
     * @return true if the algorithm can encrypt
     */
    public boolean isEncryptionCapable() {
        return encryptionCapable;
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/algorithm/PublicKeyAlgorithm.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.algorithm

import org.bouncycastle.bcpg.PublicKeyUtils

/**
 * Enumeration of public key algorithms as defined in RFC4880, RFC9580, Persistent Symmetric Keys.
 *
 * @see [RFC4880: Public-Key Algorithms](https://tools.ietf.org/html/rfc4880#section-9.1)
 * @see
 *   [RFC9580: Public-Key Algorithms](https://www.rfc-editor.org/rfc/rfc9580.html#name-public-key-algorithms)
 * @see
 *   [Persistent Symmetric Keys in OpenPGP](https://www.ietf.org/archive/id/draft-ietf-openpgp-persistent-symmetric-keys-01.html#name-persistent-symmetric-key-al)
 */
enum class PublicKeyAlgorithm(val algorithmId: Int) {

    // RFC4880

    /** RSA capable of encryption and signatures. */
    RSA_GENERAL(1),

    /**
     * RSA with usage encryption.
     *
     * @deprecated see [Deprecation notice](https://tools.ietf.org/html/rfc4880#section-13.5)
     */
    @Deprecated("RSA_ENCRYPT is deprecated in favor of RSA_GENERAL", ReplaceWith("RSA_GENERAL"))
    RSA_ENCRYPT(2),

    /**
     * RSA with usage of creating signatures.
     *
     * @deprecated see [Deprecation notice](https://tools.ietf.org/html/rfc4880#section-13.5)
     */
    @Deprecated("RSA_SIGN is deprecated in favor of RSA_GENERAL", ReplaceWith("RSA_GENERAL"))
    RSA_SIGN(3),

    /** ElGamal with usage encryption. */
    ELGAMAL_ENCRYPT(16),

    /** Digital Signature Algorithm. */
    DSA(17),

    /** Elliptic Curve Diffie-Hellman. */
    ECDH(18),

    /** Elliptic Curve Digital Signature Algorithm. */
    ECDSA(19),

    /**
     * ElGamal General.
     *
     * @deprecated see [Deprecation notice](https://tools.ietf.org/html/rfc4880#section-13.8)
     */
    @Deprecated("ElGamal is deprecated") ELGAMAL_GENERAL(20),

    /** Diffie-Hellman key exchange algorithm. */
    DIFFIE_HELLMAN(21),

    /** Digital Signature Algorithm based on twisted Edwards Curves. */
    EDDSA_LEGACY(22),

    // RFC9580

    /** X25519 encryption algorithm. */
    X25519(25),

    /** X448 encryption algorithm. */
    X448(26),

    /** Ed25519 signature algorithm. */
    ED25519(27),

    /** Ed448 signature algorithm. */
    ED448(28),

    // Persistent Symmetric Keys in OpenPGP

    /**
     * AEAD can be used as a persistent key symmetric encryption algorithm for message encryption.
     *
     * @see
     *   [Persistent Symmetric Keys in OpenPGP](https://datatracker.ietf.org/doc/draft-ietf-openpgp-persistent-symmetric-keys/)
     */
    AEAD(128) {
        override val signingCapable = false
        override val encryptionCapable = true
    },

    /**
     * HMAC can be used as a persistent key symmetric signing algorithm for message signing.
     *
     * @see
     *   [Persistent Symmetric Keys in OpenPGP](https://datatracker.ietf.org/doc/draft-ietf-openpgp-persistent-symmetric-keys/)
     */
    HMAC(129) {
        override val signingCapable = true
        override val encryptionCapable = false
    };

    open val signingCapable: Boolean = PublicKeyUtils.isSigningAlgorithm(algorithmId)
    open val encryptionCapable: Boolean = PublicKeyUtils.isEncryptionAlgorithm(algorithmId)

    fun isSigningCapable(): Boolean = signingCapable

    fun isEncryptionCapable(): Boolean = encryptionCapable

    companion object {
        @JvmStatic
        fun fromId(id: Int): PublicKeyAlgorithm? {
            return values().firstOrNull { it.algorithmId == id }
        }

        @JvmStatic
        fun requireFromId(id: Int): PublicKeyAlgorithm {
            return fromId(id)
                ?: throw NoSuchElementException("No PublicKeyAlgorithm found for id $id")
        }
    }
}
